### DISCUSSION TITLE
inoremap issue / key remapping priority

### DISCUSSION THREAD

**Discussion by sourcevault** [NeoVintageous/NeoVintageous#920] [Q&A] [answered]
Hi,

I am having some trouble with remapping the `pageup` and `pagedown` keys:

https://github.com/NeoVintageous/NeoVintageous/assets/20347647/0d561768-40b5-406c-b4e6-5113054e1c23

Basically the following code is not working in my system :

```vim
inoremap <PageUp> 5k
inoremap <PageDown> 5j
```

and trying to use sublime text's keybinding to do the remapping breaks the same keybinding somewhere else :

```vim
vnoremap <PageUp> 5k
vnoremap <PageDown> 5j

nnoremap <PageUp> 5k
nnoremap <PageDown> 5j
```

I understand why it would break it, but dont know how to resolve the problem.

**Comment by gerardroche** [comment, marked as answer]
Very few insert mode mappings are implemented. There is an open issue for them https://github.com/NeoVintageous/NeoVintageous/issues/837

**Reply by sourcevault** [reply]
Hi @gerardroche, thank you for your quick reply 🙏.

Is there a workaround using sublime's normal key mapping. 

It seems it works as long as the default mapping is not disturbed. 

As in, only apply sublime keybinding ( in `User\Default.sublime-keymap` file ) in **certain neovintageous context** - in this case, only use the sublime keybinding in neovintageous insert mode.

**Comment by gerardroche** [comment]
You could try define the following keymap:

```
    {"keys": ["pagedown"],  "command": "nv_feed_key", "args": {"key": "<pagedown>"},   "context": [{"key": "vi_insert_mode_aware"},  {"key": "nv_handle_key", "operand": "<PageDown>"}]},
    {"keys": ["pageup"],  "command": "nv_feed_key", "args": {"key": "<pageup>"},   "context": [{"key": "vi_insert_mode_aware"},  {"key": "nv_handle_key", "operand": "<PageUp>"}]},
```

The main reason I haven't implemented insert mode keys is because they tend to interfere with the normal operations in sublime. For example widgets like search, inputs, and overlays. 

If you define the keymaps above, the pagedown and pageup will probably not work inside widgets. It probably won't matter for these two keys because you probably never use them inside widgets.  But many other keys are more complicated.

**Reply by sourcevault** [reply]
Thank you for that answer 🙏, could not work without it.

I made it work today 🍕 .

I have a question for you that I am dying to know your answer.

Do you make extensive use of the following navigation keys :

```
<C-f>
<C-u>
<C-d>
<C-b>
```

What about these :

```
Shift+l
Shift+m
Shift+h
```
They seem really strange to me to be used for navigation, as they require multiple key-press.

I also do not know where my cursor will land.

With `5j` I know exactly where my cursor will be.

Also these keybinding conflicts with many useful ones like :

```
<C-f> # sublime text find.
<C-d> # sublime text multiple cursor.
```

and the one for multiple cursors `<C-n>` conflicts with a useful sublime text command - `new_file`.

Thanks.

**Reply by gerardroche** [reply]
Yes. I use `CTRL-d` and `CTRL-u` all the time. It's a quick way to scroll pages. I don't use `CTRL-f` and `CTRL-b`. The latter two scroll a half the visible lines so I think it's easier to see all the code as you scroll.  

I don't use those keys very much.  There are so many vim commands. Sometimes you forget about commands. But over time you add more and more commands into your toolbox. For example, I kinda forgot about the `CTRL-f` and `CTRL-b` keys, but now you've made me remember them I might start using them. Time will tell.

I think people new to vim sometimes try to learn everything all at once. That's not possible. Start with the basics to get back to your normal productivity level, and then over time add more and more commands into the toolbox.


You can tell nv to ignore keys, in which case they will fallback to their ST functionality. For example, here is mine:

```js
	"vintageous_handle_keys":
	{
		"<C-f12>": false,
		"<S-f1>": false,
		"<S-tab>": false,
		"<f1>": false,
		"i_<C-u>": false,
	},
```  

Read my [blog post on the neovintageous key handler](https://blog.gerardroche.com/2022/09/22/neovintageous-key-handler/) and [my Sublime Text setup](https://blog.gerardroche.com/2023/06/12/my-sublime-text-setup/).

**Reply by sourcevault** [reply]
> I think people new to vim sometimes try to learn everything all at once. That's not possible. Start with the basics to get back to your normal productivity level, and then over time add more and more commands into the toolbox.

You are correct in saying this, but navigation is one of those core functionality that you have to get right from the start.

I think the problem is best described using a keyboard layout analogy.

If you have some missing keys - like 'q' of your keyboard - it is really frustrating, even if you use 'q' less that 0 times every 1000 clicks.

So 99% of people are unable to adapt to a different keyboard layout or a different editor like vim.

Neovintageous is the first tool I came across in over a decade of trying to switch to vim that makes it possible for normie users.

**Reply by gerardroche** [reply]
Yes. I've never been able to switch to Vim. But I wonder now if NeoVintageous or any other emulator was taken away from me, would I then switch to Vim. I think I would end up using both Vim and an IDE together. Over time I would probably end up dropping the IDE. 

I think there are a few reasons. Using the mouse is just more easier for some things. People say use gvim, but it's not the same. 

I think search and search & replace across a large project is a big learning curve in Vim. On a single buffer search `/{pattern}` and substitute `:s/{pattern}/{string}/` are great. [`smartcase` searching](https://blog.gerardroche.com/2020/12/11/the-curious-case-of-vim-search/) is great. But across many files it's more difficult in Vim. At least, there is a big learning curve. I suspect once you learn it well, it's probably better.

Another obstacle with Vim is that you need to setup lots of plugins and configure it to get to a useable state. It's unreasonable to expect new users to be able to setup Vim first time.  The new Vim user first experience is not great. Instead of learning commands, you typically have to start learning how to setup Vim.

The first thing I tell people is press `<Esc>` to normal mode. If you're not sure what mode your in press `<Esc>` several times, now you're in normal mode. `i` to insert mode. When you're in insert mode, you're editor mostly function normally. So if your confused. Press `<Esc>` several times. What mode will you be in? Normal. Now press `i` to enter insert mode and you can mostly edit normally.

Beyond that. You just need to know how to save a file `:w` and exit `:q`.

When you think about it. When you're in insert mode, you can use the arrow keys to navigate. So in insert mode, vim works pretty much like any other editor.

I think users get too bogged down in trying to learn tons of motions before learning the basic principles that will get you out of a jam. If you get frustrated because vim is beeping at you, press `<Esc>` several times. Now enter insert mode and get done what you need to do.  Then learn a few motions at a time.

But it's easier said than done isn't it. There a lots of obstacles that forgotten about. Like how you need to generate ctags for jump to declarations to work. That's a whole thing you have learn. Or how you need install and learn [NerdTree](https://github.com/preservim/nerdtree) for side bar navigation. This isn't something you need to learn in Sublime+Neo. Some of the features are available, but you don't *need* to use them. In Vim you do need them.

I think emulators are a fair middle ground. And the fact every editor and IDE has a vim emulation plugin, shows how there is a market for a native vim + sublime/ide styled editor. Many have tried, but I've yet to see a good one.

**Reply by sourcevault** [reply]
> Yes. I've never been able to switch to Vim. But I wonder now if NeoVintageous or any other emulator was taken away from me, would I then switch to Vim. I think I would end up using both Vim and an IDE together. Over time I would probably end up dropping the IDE.

> Another obstacle with Vim is that you need to setup lots of plugins and configure it to get to a useable state. It's unreasonable to expect new users to be able to setup Vim first time. The new Vim user first experience is not great. Instead of learning commands, you typically have to start learning how to setup Vim.

I tried using neovim for a few days and gave up trying to configure it.

Mind you, I have been able to configure arch linux on a laptop, so I can say with confidence that setting up neovim is harder than installing arch linux, and that is saying something.

> I think there are a few reasons. Using the mouse is just more easier for some things. People say use gvim, but it's not the same.

Mouse is really important for discoverability.

> I think users get too bogged down in trying to learn tons of motions before learning the basic principles that will get you out of a jam. If you get frustrated because vim is beeping at you, press <Esc> several times. Now enter insert mode and get done what you need to do. Then learn a few motions at a time.

Learning a new natural language is considered difficult, but it shouldn't be, most people are not interested in the abstract grammar or alphabets intrigue of a particular language. 

Languages are easier to soak up if you associate it with a excellent story that generates strong emotions. In the case of English it is things like Harry Potter or Lord of the Rings, in the case of French it is things like Les Misérables.

This is true for vim too, people care about how vim is going to help them accomplish certain tasks quickly, people are waiting for the Les Misérable for Vim.

For me, I did not like how I was navigating within a single line of code. I would use `<home>`,`<end>`,`<left>` and `<right>`. I remember vintage mode in sublime text had a better way using `b` and `w`.

So while doing research on sublime text vintage mode, I came across your project.

**Reply by gerardroche** [reply]
I've been thinking about it. After learning `<Esc>` to normal mode, `i` insert mode (it's amazing how many new vim users don't grok this fundamental), I think one of the next major obstacles is copy and paste, but users consistently focus on motions. Most of the motions are about productivity.

For example, as an exercise try working on some code in only in insert mode. You'll have to the use the arrows to navigate around. What task is near impossible to complete? It's not navigating to specific line, or using:

- `w` forward one word 
- `b` back one word
- `^` to move to the beginning of a line
- `$` to the end of line
- `gg` to beginning of file
- `f{char}` forward to char
- `F{char}` back to char
- `t{char}` forward till char
- `T{char}` back till char
- `G` end of file. 

It's not learning about counts `{count}f{char`, `{count}w`, etc.

New vim users are often hit with all these and more.

It's copy and paste. It took me ages to learn how to efficiently copy (yank) and paste. Learning text objects is a big learning curve, but you need to really grok basic yank and paste. I think that's when you stop using the mouse so much.

I find navigating code with 

- `CTRL-]` jump to definition (`CTRL-w ]` will split it)
- `CTRL-o` go back (cursor position history)
- `CTRL-i` go forward (cursor position history)

I use those a lot.

I also use window navigations key a lot:

- `CTRL-w h`
- `CTRL-w j`
- `CTRL-w k`
- `CTRL-w l`

Splits

- `CTRL-w s`
- `CTRL-w v`
- `CTRL-w n`

Window layout when you have a split (two view side by side)

- `CTRL-w H`
- `CTRL-w J`
- `CTRL-w K`
- `CTRL-w L`

I'll bet not many know those ones. Super useful.

Resize all views equally.

- `CTRL-w =`

I use that a lot too when my windows are messed up after I resize thing with

-  `CTRL-w |` make current window max width
-  `CTRL-w _` make current window max height

Now `CTRL-w =` to resize them all equally.

- `CTRL-w q` close current view
- `CTRL-w o` close all except current

I use the `CTRL-w` commands a lot. They were all the first commands I implemented when I forked Vintageous.

I could keep going. But it has taken me years to learn all these things. I still think copy and paste is big one for new users because in programming you do so much of it. I used to use the mouse a lot and there was no one point when I stopped. I've just gradually used it less and less.

Visual line mode is useful for selecting block to yank. For example, entering visual line mode, `V`, and then using the arrow keys is the simple way to select a block. It may not be the most efficient way, but it's certainly simple. Then `y` yanks it and `p` or `P` will paste it.

Text objects start to make things more efficient: `V}` and `V{` or `v}` and `v{`.

You can get a lot of mileage from those.

As I'm writing, I'm thinking of more commands I use a lot.

Navigating symbols under the cursor:

- `*` forward to next symbol under cursor
- `#` backward to previous symbol under cursor

That's handy because it also highlights the occurrences.

- `%` to next brace/bracket

I think the lightbulb moment for me was groking how any motion can be used with any operator.



- `w` move a word forward
- `yw` yank word  
- `dw` delete word
- `cw` delete word
- `yiw` yank inner word  
- `diw` delete inner word
- `ciw` delete inner word
- `yaw` yank around word  
- `daw` delete around word
- `caw` delete around word

Now just change the motion. 

- `}` move a paragraph forward
- `y}` yank paragraph  
- `d}` delete paragraph
- `c}` delete paragraph
- `yi}` yank inner paragraph (only works inside braced block `{...}`)  
- `di}` delete inner paragraph (only works inside braced block `{...}`)
- `ci}` delete inner paragraph (only works inside braced block `{...}`)
- `ya}` yank around paragraph (only works inside braced block `{...}`)  
- `da}` delete around paragraph (only works inside braced block `{...}`)
- `ca}` delete around paragraph (only works inside braced block `{...}`)

So for example

- `>}` indents a paragraph
- `>i}` indents an inner paragraph
- `>a}` indents around a paragraph

deindent

- `<}` indents a paragraph
- `<i}` indents an inner paragraph
- `<a}` indents around a paragraph

At a certain point when you learn a new motion, you actually learn tons of new commands.

You also use `v` as an operator e.g. `vw`, `viw`, `vaw`. Now just change the motion.

More and more of the commands I use every day are occurring to me. I'll have to extract all this talking into a blog post.

**Comment by sourcevault** [comment]
Alright this was a really bad idea 😢 .

It broke `Ctrl + PageUp` and `Ctrl + PageDown` which is used to switch between tabs.

Basically anything that you have mapped via sublime will break.

I have this piece of code that allows me to assign a key binding to move by **X** number of lines:

```python
class MoveByLineCommand(sublime_plugin.TextCommand):
    def run(self, edit, forward=True, lines=5):
        region = self.view.visible_region()

        height = lines

        if not forward:
            height = -height

        (row, col) = self.view.rowcol(self.view.sel()[0].begin())
        target = self.view.text_point(row + height, col)

        self.view.sel().clear()
        self.view.sel().add(sublime.Region(target))
        self.view.show(target)
```

and I was using it in neovintageous like this :

```vim
inoremap <PageUp> :MoveByLine forward=false lines=5<CR>
inoremap <PageDown> :MoveByLine forward=true lines=5<CR>
```

It should have been used in the sublime keybinding file like below, (and kept out of neovintageous ) :

```json 
{ "keys": ["pageup"], "command": "move_by_line", "args": {"forward": false}},
{ "keys": ["pagedown"], "command": "move_by_line", "args": {"forward": true}},
```

**Reply by gerardroche** [reply]
Yes. The insert mode mappings are problematic. This is why they are not implemented. They break too much of the editor basic functionality. Neo would have to replicate all that functionality, or at least handle proxying the command to st if the command is a noop. It's a tough one.
